index.js

```diff
@@ -33,28 +33,61 @@ function TinySDF(fontSize, buffer, radius, cutoff, fontFamily, fontWeight) {
     this.middle = Math.round((size / 2) * (navigator.userAgent.indexOf('Gecko/') >= 0 ? 1.2 : 1));
 }
 
-TinySDF.prototype.draw = function (char) {
+TinySDF.prototype._draw = function (char, getMetrics) {
     this.ctx.clearRect(0, 0, this.size, this.size);
     this.ctx.fillText(char, this.buffer, this.middle);
+    var textMetrics = this.ctx.measureText(char);
+    // Older browsers only expose the glyph width
+    // This is enough for basic layout with all glyphs using the same fixed size
+    var metrics = {
+      width: this.size,
+      height: this.size,
+      left: 0,
+      top: -this.size / 3,
+      advance: textMetrics.width
+    };
+
+    var imgData;
+    // If the browser supports bounding box metrics, we can generate a smaller
+    // SDF. This is a significant performance win.
+    if (getMetrics && textMetrics.actualBoundingBoxLeft !== undefined) {
+      metrics.width = Math.ceil(textMetrics.actualBoundingBoxRight - textMetrics.actualBoundingBoxLeft);
+      metrics.height = Math.ceil(textMetrics.actualBoundingBoxAscent + textMetrics.actualBoundingBoxDescent);
+      metrics.top = Math.floor(textMetrics.actualBoundingBoxAscent) - this.middle;
+
+      // The integer/pixel part of the top alignment is encoded in metrics.top
+      // The remainder is implicitly encoded in the rasterization
+      var imgTop = this.middle - Math.ceil(textMetrics.actualBoundingBoxAscent) - this.buffer;
+      imgData = this.ctx.getImageData(0, imgTop, metrics.width + 2 * this.buffer, metrics.height + 2 * this.buffer);
+    } else {
+      imgData = this.ctx.getImageData(0, 0, this.size, this.size);
+    }
 
-    var imgData = this.ctx.getImageData(0, 0, this.size, this.size);
-    var alphaChannel = new Uint8ClampedArray(this.size * this.size);
+    var alphaChannel = new Uint8ClampedArray(imgData.width * imgData.height);
 
-    for (var i = 0; i < this.size * this.size; i++) {
+    for (var i = 0; i < imgData.width * imgData.height; i++) {
         var a = imgData.data[i * 4 + 3] / 255; // alpha value
         this.gridOuter[i] = a === 1 ? 0 : a === 0 ? INF : Math.pow(Math.max(0, 0.5 - a), 2);
         this.gridInner[i] = a === 1 ? INF : a === 0 ? 0 : Math.pow(Math.max(0, a - 0.5), 2);
     }
 
-    edt(this.gridOuter, this.size, this.size, this.f, this.v, this.z);
-    edt(this.gridInner, this.size, this.size, this.f, this.v, this.z);
+    edt(this.gridOuter, imgData.width, imgData.height, this.f, this.v, this.z);
+    edt(this.gridInner, imgData.width, imgData.height, this.f, this.v, this.z);
 
-    for (i = 0; i < this.size * this.size; i++) {
+    for (i = 0; i < imgData.width * imgData.height; i++) {
         var d = Math.sqrt(this.gridOuter[i]) - Math.sqrt(this.gridInner[i]);
         alphaChannel[i] = Math.round(255 - 255 * (d / this.radius + this.cutoff));
     }
 
-    return alphaChannel;
+    return { alphaChannel, metrics };
```

Comment:
Nit: let's keep this version ES5 for consistency, and do ES6 migration in a separate PR later.

Reply:
Should we remove the `metrics` nesting and return an object with `alphaChannel`, `width`, etc.? Also, perhaps `alphaChannel` sounds a bit too verbose as a property — maybe something short like `data` would be nicer?

Reply:
Oops, didn't even realize that was ES6 notation.

Switched to `data`.